File: 0015-3sum/0015-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {
        vector<vector<int>> output;

        sort(nums.begin(), nums.end());
        int sum;
        int l, r;
        for(int i = 0; i < nums.size() - 2; i++) {
            if(nums[i] > 0) break;
            if((i > 0) && (nums[i] == nums[i-1])) continue;

            l = i + 1;
            r = nums.size() - 1;

            while(l < r) {
                sum = nums[i] + nums[l] + nums[r];
                if(sum > 0) {
                    r--;
                }
                else if(sum < 0) {
                    l++;
                }
                else {
                    output.push_back({nums[i], nums[l], nums[r]});

                    while((l < r) && (nums[l] == nums[l+1])) l++;
                    while((l < r) && (nums[r] == nums[r-1])) r--;

                    l++;
                    r--;
                }
            }
        }
        return output;
    }
};

File: 0015-3sum/0015-3sum.py
class Solution:
    def threeSum(self, nums: list[int]) -> list[list[int]]:
        size = len(nums)
        nums.sort()
        output = []
        for i in range(0, size):

            if nums[i] > 0:
                break
                
            if i > 0 and nums[i] == nums[i - 1]:
                continue

            l = i + 1
            r = size - 1

            while l < r:

                total = nums[i] + nums[l] + nums[r]
                if total < 0:
                    l += 1
                
                elif total > 0:
                    r -= 1
                
                else:
                    output.append([nums[i], nums[l], nums[r]])
                    while l < r and nums[l] == nums[l + 1]:
                        l += 1
                    
                    while l < r and nums[r] == nums[r - 1]:
                        r -= 1
                    
                    l += 1
                    r -= 1
        
        return output


        